### DISCUSSION TITLE
Diesel boxed query with length_aware_paginator

### DISCUSSION THREAD

**Discussion by SarcevicLuka** [diesel-rs/diesel#3760] [Q&A]
Hello, I have a problem with paginating my boxed query.

[Project link](https://github.com/SarcevicLuka/diplomski-rad-backend/tree/master/bins/monolith/src/application/post/get) to see for yourself in **src/application/post/get** directory (It is commented out)

This is my function, and I have three tables. A Post table that contains the watch review, a foreign keys to a user who created the post and a watch that holds tha watch data. I want to create a query where I get the last week posts and paginate it. The paginated response should contain the post user and watch data (all structs shown below the function in the To reproduce section).

    async fn get_users_posts_paginated(
        &self,
        user_id: &str,
        attributes: GetPostsAttributes
    ) -> Result<Response<(Option<Post>, Option<User>, Option<Watch>)>, Error> {

        let mut conn = self.pg_pool.connection()?;

        let mut query = 
            posts::table
                .left_join(users::table)
                .left_join(watches::table)
                .select((
                    Option::<Post>::as_select(),
                    Option::<User>::as_select(),
                    Option::<Watch>::as_select()
                ))
                .into_boxed();
                
        query = query
            .filter(posts::user_id.eq(user_id))
            .order_by(posts::created_at.desc());

        query
            .page(attributes.page)
            .per_page(attributes.per_page)
            .paginate(&mut conn)
            .map_err(Error::from)
    }

However I get this error on paginate(&mut conn) function:

```
'error[E0277]: the trait bound '(diesel::expression::select_by::SelectBy<std::option::Option<Post>, Pg>, diesel::expression::select_by::SelectBy<std::option::Option<User>, Pg>, diesel::expression::select_by::SelectBy<std::option::Option<Watch>, Pg>): diesel::util::TupleSize' is not satisfied
  --> bins\monolith\src\application\post\get\infrastructure.rs:91:14
   |
91 |             .paginate(&mut conn)
   |              ^^^^^^^^ the trait 'diesel::util::TupleSize' is not implemented for '(diesel::expression::select_by::SelectBy<std::option::Option<Post>, Pg>, diesel::expression::select_by::SelectBy<std::option::Option<User>, Pg>, diesel::expression::select_by::SelectBy<std::option::Option<Watch>, Pg>)'
   |
  --> C:\Users\Korisnik\.cargo\registry\src\github.com-1ecc6299db9ec823\length_aware_paginator-1.0.0\src\traits.rs:89:15
   |
89 |         Self: LoadQuery<'query, PgConnection, (U, i64)>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in 'length_aware_paginator::traits::<impl PaginatedQuery<T>>::paginate
```
 
The paginate is a function from _length_aware_paginator_ crate and looks like this:
```
pub fn paginate<'query, U>(self, conn: &mut PgConnection) -> QueryResult<Response<U>>
    where
        Self: LoadQuery<'query, PgConnection, (U, i64)>,
    {
        let page = self.page;
        let per_page = self.per_page;
        let results = self.load::<(U, i64)>(conn)?;
        let total = results.get(0).map(|x| x.1).unwrap_or(0);
        let data = results.into_iter().map(|x| x.0).collect();
        let mut last_page = 1;

        if total > 0 {
            last_page = (total as f64 / per_page as f64).ceil() as i64;
        }

        Ok(Response {
            page,
            per_page,
            total,
            last_page,
            data,
        })
    }
```
Right here I see a tuple with i64, but I do not understand why it comes up only here and not in the example that works for me, what i64 does it want from my querry.

**To reproduce:**

_cargo.toml_
```
# Helpers
serde = "1.0.157"
serde_json = "1.0.94"

# Http
actix-web = "4.3.1"
actix-cors = "0.6.4"

# Database
diesel = { version = "2.0.3", features = [
  "chrono",
  "postgres",
  "r2d2",
  "serde_json",
  "64-column-tables",
] }
r2d2 = "0.8.10"
length_aware_paginator = "1.0.0"
```

_schema.rs_
```
diesel::table! {
    posts (id) {
        #[max_length = 36]
        id -> Varchar,
        #[max_length = 36]
        user_id -> Varchar,
        #[max_length = 255]
        watch_id -> Varchar,
        review -> Text,
        score -> Int4,
        num_of_likes -> Int4,
        num_of_comments -> Int4,
        created_at -> Timestamptz,
        updated_at -> Timestamptz,
    }
}
diesel::table! {
    users (id) {
        #[max_length = 36]
        id -> Varchar,
        #[max_length = 255]
        email -> Varchar,
        #[max_length = 255]
        first_name -> Varchar,
        #[max_length = 255]
        last_name -> Varchar,
        #[max_length = 255]
        password -> Varchar,
        avatar -> Bytea,
        created_at -> Timestamptz,
        updated_at -> Timestamptz,
    }
}
diesel::table! {
    watches (id) {
        #[max_length = 36]
        id -> Varchar,
        #[max_length = 255]
        brand -> Varchar,
        #[max_length = 255]
        model -> Varchar,
        diameter -> Int4,
        lug_width -> Int4,
        #[max_length = 255]
        case_material -> Varchar,
        #[max_length = 255]
        mechanism_model -> Varchar,
        created_at -> Timestamptz,
        updated_at -> Timestamptz,
    }
}
```

_data.rs_ that holds the request and response data
```
#[derive(Clone, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct GetPostsAttributes {
    pub page: Option<i64>,
    pub per_page: Option<i64>,
    pub sort: Option<String>,
    pub name: Option<String>,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PaginatedPostsResponse {
    pub page: i64,
    pub per_page: i64,
    pub total: i64,
    pub last_page: i64,
    pub data: Vec<PostListResponse>,
}

impl From<Response<(Option<Post>, Option<User>, Option<Watch>)>> for PaginatedPostsResponse {
    fn from(source: Response<(Option<Post>, Option<User>, Option<Watch>)>) -> Self {
        Self {
            page: source.page,
            per_page: source.per_page,
            total: source.total,
            last_page: source.last_page,
            data: source
                .data
                .into_iter()
                .map(PostListResponse::from)
                .collect::<Vec<PostListResponse>>(),
        }
    }
}

/// Struct that holds posts for the frontend
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PostListResponse {
    pub post: Post,
    pub creator: DisplayUser,
    pub watch_data: Watch,
}

impl From<(Option<Post>, Option<User>, Option<Watch>)> for PostListResponse {
    fn from(source: (Option<Post>, Option<User>, Option<Watch>)) -> Self {
        let (post, creator, watch_data) = source;
        PostListResponse { 
            post: post.unwrap(),
            creator: DisplayUser::from(creator.unwrap()),
            watch_data: watch_data.unwrap()
        }
    }
}

/// Struct that holds users pagination attributes
#[derive(Clone, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct GetPostsAttributes {
    pub page: Option<i64>,
    pub per_page: Option<i64>,
    pub sort: Option<String>,
    pub name: Option<String>,
}
```

_domain.rs_ that holds the main logic
```
async fn get_users_posts_paginated(
        &self,
        user_id: &str,
        attibutes: GetPostsAttributes
    ) -> Result<PaginatedPostsResponse, Error> {
        self
            .repository
            .get_users_posts_paginated(user_id, attibutes)
            .await
            .map(PaginatedPostsResponse::from)
    }
```

I think it is something with the left_join or the select statements because in the error message it says LoadQuery<'query, PgConnection, (U, i64)>. I don't know why I get the i64 in the tuple.

When I do a similar query like this:
```
    async fn get_users_posts_paginated(
        &self,
        user_id: &str,
        attributes: GetPostsAttributes
    ) -> Result<Response<(Post, i64)>, Error> {

        let mut conn = self.pg_pool.connection()?;

        let mut query = 
            posts::table
                .left_join(post_likes_count)
                .group_by(posts::id)
                .select((posts::all_columns, sql::<diesel::sql_types::BigInt>("COUNT(post_likes.id) as num_likes")))
                .into_boxed();
                
        query = query.filter(posts::user_id.eq(user_id));

        query
            .page(attributes.page)
            .per_page(attributes.per_page)
            .paginate(&mut conn)
            .map_err(Error::from)
    }
```
I don't get any errors. And I have the i64 in the return type but that is the count from the query. 

Am I doing the left_join or select wrong?

**Comment by weiznich** [comment]
It's unfortunately hard to help you based on the provided information. Try to:

* Format your post in a way that's readable for others
* Minimize your example by removing unneeded types/…
* Include all the information required to reproduce the error message. That includes the relevant parts of your schema.rs file, definitions for all types not belonging to diesel or rust-std and imports for all types that belong to rust-std or diesel.

**Reply by SarcevicLuka** [reply]
I think it should be clearer now. I hope it has all the info you need, if not ask for more details I will provide as much as you need

**Reply by weiznich** [reply]
>  I hope it has all the info you need, if not ask for more details I will provide as much as you need

No it doesn't have all the information I've asked for.  It's especially missing information about the `paginate` method the error is coming from. This method is not provided by diesel, so its totally unclear how it's defined. Please keep in mind that it's really frustrating to ask for these things in the first place already, so really try to include everything that's required to reproduce your problem. After all it makes it harder for people to help you, which means its just more likly that they won't brother at all. 
Ideally you should provide a complete minimal project that shows your problem and where someone can just to a `cargo check` to get exactly the same error message in your question.

**Reply by SarcevicLuka** [reply]
Sorry for the incovenience I am preety new to this stuff.  Ill update the question.

**Reply by weiznich** [reply]
Thanks for updating the post. It seems like that this is an issue with `length_aware_paginator`, as the query itself compiles if I replace the last part just with a `RunQueryDsl::load()`. I'm not familiar with the implementation of that crate, therefore it's hard to say what's exactly the problem there. You might want to reach out in their support channels.

**Reply by SarcevicLuka** [reply]
Thank you for your time. I want to ask one more thing. Is there a difference in 
```
posts::table
            .left_join(users::table)
            .left_join(watches::table)
            .select((
                Option::<Post>::as_select(),
                Option::<User>::as_select(),
                Option::<Watch>::as_select()
            ))
            .into_boxed();
```
and
```
 posts::table
        .left_join(post_likes_count)
        .group_by(posts::id)
        .select((posts::all_columns, sql::<diesel::sql_types::BigInt>("COUNT(post_likes.id) as num_likes")))
        .into_boxed();
```
in the data shape these queries hold?
Because in the firstone I fetch the data with _as_select()_ and in the on that works I fetch it with _all_columns_. Is there a difference between these two?

When I fetch only posts with _all_columns_ it works fine
```
let mut query = 
            posts::table
                .left_join(users::table)
                .left_join(watches::table)
                .select((
                    posts::all_columns,
                    Option::<User>::as_select(),
                    Option::<Watch>::as_select()
                ))
                .into_boxed();
```


Furthermore the error also gives me this
```
   = note: required for `(Post, std::option::Option<User>, std::option::Option<Watch>)` to implement `StaticallySizedRow<((Text, Text, Text, Text, Integer, Integer, Integer, Timestamptz, Timestamptz), diesel::expression::select_by::SelectBy<std::option::Option<User>, Pg>, diesel::expression::select_by::SelectBy<std::option::Option<Watch>, Pg>), Pg>`
   = note: required for `((Post, std::option::Option<User>, std::option::Option<Watch>), i64)` to implement `FromStaticSqlRow<(((Text, Text, Text, Text, Integer, Integer, Integer, Timestamptz, Timestamptz), diesel::expression::select_by::SelectBy<std::option::Option<User>, Pg>, diesel::expression::select_by::SelectBy<std::option::Option<Watch>, Pg>), BigInt), Pg>`
   = note: required for `((Post, std::option::Option<User>, std::option::Option<Watch>), i64)` to implement `Queryable<(((Text, Text, Text, Text, Integer, Integer, Integer, Timestamptz, Timestamptz), diesel::expression::select_by::SelectBy<std::option::Option<User>, Pg>, diesel::expression::select_by::SelectBy<std::option::Option<Watch>, Pg>), BigInt), Pg>`
   = note: required for `((Post, std::option::Option<User>, std::option::Option<Watch>), i64)` to implement `FromSqlRow<(((Text, Text, Text, Text, Integer, Integer, Integer, Timestamptz, Timestamptz), diesel::expression::select_by::SelectBy<std::option::Option<User>, Pg>, diesel::expression::select_by::SelectBy<std::option::Option<Watch>, Pg>), BigInt), Pg>`

```
I just wanted to ask you, because these are diesel traits could this just be a side effect of the first error or do I need to implement them. I am sure Queryable is implemented so I am suspicious.

**Comment by SarcevicLuka** [comment]
### **I mannaged to solve the problem.** 

The problem was in the select statement and the return type of the function.
- The return type should be a tuple of 
`(Post, Option<User>, Option<Watch>)`
- the select statement should be removed completely
```
let mut query = 
            posts::table
                .left_join(users::table)
                .left_join(watches::table)
                .into_boxed();
```
And now it works fine.

I also created a struct for cleaner code
```
#[derive(Queryable, PartialEq, Debug)]
pub struct CombinedData {
    post: Post,
    user: Option<User>,
    watch: Option<Watch>,
}
```
which is now the return type of the function
```
async fn get_feed_best_reviewed_posts_paginated(
        &self,
        attributes: GetPostsAttributes
    ) -> Result<Response<CombinedData>, Error> {
```

I am not sure why the other one did not work. My guess is that my select statement was incorrect because I read that if there is no select statement in a join, diesel will select everything automatically. And that selected data is nicely queried in the CombinedData struct and passed to the _paginate()_ method of the _length_aware_paginator_.

The User and Watch are optional because I get an error when they are not
```
.paginate(&mut conn)
    |              ^^^^^^^^ the trait `FromStaticSqlRow<((Text, Text, Text, Text, Integer, Integer, Integer, Timestamptz, Timestamptz), Nullable<(Text, Text, Text, Text, Text, diesel::sql_types::Binary, Timestamptz, Timestamptz)>, Nullable<(Text, Text, Text, Integer, Integer, Text, Text, Timestamptz, Timestamptz)>), Pg>` is not implemented for `(Post, User, Watch)`
```
My guess is also the join and select combination, because the Post table is the main table (it holds foreign keys of other tables) it is selected, but two other are optional (not sure why).

**Reply by weiznich** [reply]
> My guess is also the join and select combination, because the Post table is the main table (it holds foreign keys of other tables) it is selected, but two other are optional (not sure why).

`User` and `Watch` needs to be optional as a `left_join()` might return null values there.
